plugins/modules/packaging/language/cpanm.py

```diff
@@ -109,105 +142,114 @@
+    ModuleHelper, CmdMixin, ArgFormat, ModuleHelperException
+)
+
+
+class CPANMinus(CmdMixin, ModuleHelper):
+    output_params = ['name', 'version']
+    module = dict(
+        argument_spec=dict(
+            name=dict(type='str', aliases=['pkg']),
+            version=dict(type='str'),
+            from_path=dict(type='path'),
+            notest=dict(type='bool', default=False),
+            locallib=dict(type='path'),
+            mirror=dict(type='str'),
+            mirror_only=dict(type='bool', default=False),
+            installdeps=dict(type='bool', default=False),
+            system_lib=dict(type='bool', default=False, aliases=['use_sudo'],
+                            removed_in_version="4.0.0", removed_from_collection="community.general"),
+            executable=dict(type='path'),
+            mode=dict(type='str', choices=['compatibility', 'new'], default='compatibility'),
+            name_check=dict(type='str')
+        ),
+        required_one_of=[('name', 'from_path')],
 
-    return cmd
-
-
-def _get_cpanm_path(module):
-    if module.params['executable']:
-        result = module.params['executable']
-    else:
-        result = module.get_bin_path('cpanm', True)
-    return result
-
-
-def main():
-    arg_spec = dict(
-        name=dict(default=None, required=False, aliases=['pkg']),
-        from_path=dict(default=None, required=False, type='path'),
-        notest=dict(default=False, type='bool'),
-        locallib=dict(default=None, required=False, type='path'),
-        mirror=dict(default=None, required=False),
-        mirror_only=dict(default=False, type='bool'),
-        installdeps=dict(default=False, type='bool'),
-        system_lib=dict(default=False, type='bool', aliases=['use_sudo']),
-        version=dict(default=None, required=False),
-        executable=dict(required=False, type='path'),
     )
-
-    module = AnsibleModule(
-        argument_spec=arg_spec,
-        required_one_of=[['name', 'from_path']],
+    command = 'cpanm'
+    command_args_formats = dict(
+        notest=dict(fmt="--notest", style=ArgFormat.BOOLEAN),
+        locallib=dict(fmt=('--local-lib', '{0}'),),
+        mirror=dict(fmt=('--mirror', '{0}'),),
+        mirror_only=dict(fmt="--mirror-only", style=ArgFormat.BOOLEAN),
+        installdeps=dict(fmt="--installdeps", style=ArgFormat.BOOLEAN),
+        system_lib=dict(fmt="--sudo", style=ArgFormat.BOOLEAN),
     )
+    check_rc = True
+
+    def __init_module__(self):
+        if self.vars.mode == "compatibility" and self.vars.name_check:
+            raise ModuleHelperException("Parameter name_check can only be used with mode=new")
+
+    def _is_package_installed(self, name, locallib, version):
+        if name is None or name.endswith('.tar.gz'):
+            return False
+
+        env = {"PERL5LIB": "%s/lib/perl5" % locallib} if locallib else {}
+        cmd = ['perl', '-le', 'use %s %s;' % (name, version)]
```

Comment:
Shouldn't this handle the case `version is None` as well?

Reply:
I think I was shuffling these things around and forgot that I left two checks in place. Thanks for spotting.